public/app/features/search/service/blugeWithFallback.ts

```diff
@@ -0,0 +1,149 @@
+import { BehaviorSubject, lastValueFrom, Observable } from 'rxjs';
+
+import {
+  CoreApp,
+  DataQueryRequest,
+  DataQueryResponse,
+  Field,
+  getDefaultTimeRange,
+  isDataFrame,
+} from '@grafana/data/src';
+import { getGrafanaDatasource } from 'app/plugins/datasource/grafana/datasource';
+import { GrafanaQuery, GrafanaQueryType } from 'app/plugins/datasource/grafana/types';
+
+import { GrafanaSearcher } from './types';
+
+type BlugeSearcherWithFallbackDeps = {
+  blugeSearcher: GrafanaSearcher;
+  sqlSearcher: GrafanaSearcher;
+};
+
+export class BlugeSearcherWithFallback implements GrafanaSearcher {
+  private blugeSearcherIsReady = new BehaviorSubject<boolean>(false);
+  private initializing = false;
+  private initialized = false;
+  private initInterval: ReturnType<typeof setInterval> | undefined;
+  private readonly firstValuePromise: Promise<unknown>;
+
+  constructor(private deps: BlugeSearcherWithFallbackDeps) {
+    this.firstValuePromise = getGrafanaDatasource().then(async (ds) => {
+      const resp = await lastValueFrom(ds.query(this.createSearchReadinessRequest({ stream: false })));
+      this.blugeSearcherIsReady.next(this.isReady(resp));
+    });
+
+    this.init();
+    // workaround.. `getGrafanaLive()` returns undefined for a while at startup
```

Comment:
interesting!  do you have queryOverLive feature flag on?

Reply:
I can reproduce this both with and without the `queryOverLive` feature flag. I tried changing the [`getGrafanaLiveSrv`](https://github.com/grafana/grafana/blob/3c6e0e8ef85048af952367751e478c08342e17b4/packages/grafana-runtime/src/services/live.ts#L123) as follows:

```ts
export const getGrafanaLiveSrv = (): GrafanaLiveSrv => {
  if (!singletonInstance) {
    initGrafanaLive();
  }
  return singletonInstance;
};
```

and got this error in console:

```
Uncaught ReferenceError: Cannot access 'StreamingFrameAction' before initialization
    at Module.StreamingFrameAction (legacyAngularInjector.ts:23:8)
    at ./.yarn/__virtual__/@grafana-runtime-virtual-f9d61a496b/1/public/app/features/live/centrifuge/service.ts (service.ts:55:11)
    at __webpack_require__ (bootstrap:19:1)
    at ./.yarn/__virtual__/@grafana-runtime-virtual-f9d61a496b/1/public/app/features/live/index.ts (utils.ts:54:8)
    at __webpack_require__ (bootstrap:19:1)
    at ./.yarn/__virtual__/@grafana-runtime-virtual-f9d61a496b/1/packages/grafana-runtime/src/services/live.ts (legacyAngularInjector.ts:23:8)
    at __webpack_require__ (bootstrap:19:1)
    at ./.yarn/__virtual__/@grafana-runtime-virtual-f9d61a496b/1/packages/grafana-runtime/src/utils/DataSourceWithBackend.ts (analytics.ts:151:8)
    at __webpack_require__ (bootstrap:19:1)
    at ./public/app/features/plugins/datasource_srv.ts (built_in_plugins.ts:149:1)
StreamingFrameAction @ legacyAngularInjector.ts:23
./.yarn/__virtual__/@grafana-runtime-virtual-f9d61a496b/1/public/app/features/live/centrifuge/service.ts @ service.ts:55
__webpack_require__ @ bootstrap:19
./.yarn/__virtual__/@grafana-runtime-virtual-f9d61a496b/1/public/app/features/live/index.ts @ utils.ts:54
__webpack_require__ @ bootstrap:19
./.yarn/__virtual__/@grafana-runtime-virtual-f9d61a496b/1/packages/grafana-runtime/src/services/live.ts @ legacyAngularInjector.ts:23
__webpack_require__ @ bootstrap:19
./.yarn/__virtual__/@grafana-runtime-virtual-f9d61a496b/1/packages/grafana-runtime/src/utils/DataSourceWithBackend.ts @ analytics.ts:151
__webpack_require__ @ bootstrap:19
./public/app/features/plugins/datasource_srv.ts @ built_in_plugins.ts:149
__webpack_require__ @ bootstrap:19
./public/app/features/plugins/all.ts @ types.ts:227
__webpack_require__ @ bootstrap:19
./public/app/features/all.ts @ time.ts:42
__webpack_require__ @ bootstrap:19
./public/app/app.ts @ loadAndInitAngularIfEnabled.ts:22
__webpack_require__ @ bootstrap:19
./public/app/index.ts @ utils.ts:309
__webpack_require__ @ bootstrap:19
__webpack_exec__ @ Utility.js:115
(anonymous) @ Utility.js:115
webpackJsonpCallback @ jsonp chunk loading:71
(anonymous) @ app.df87eb9b5dfb3231d9fe.js:1
```

I am not sure how does the initialization logic work for live - `initGrafanaLive` is called in `GrafanaCtrl` that is an "angular controller". Everything seems to work once once I started calling `initGrafanaLive` in `app/app.ts` instead of in the `GrafanaCtrl`, but there might be some unintended consequences 😬